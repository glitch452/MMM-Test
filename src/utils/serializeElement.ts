import { JSDOM } from 'jsdom';
import { JSX } from 'jsx-dom-cjs';

const WINDOW = global.window;
const DOCUMENT = global.document;
const NAVIGATOR = global.navigator;

const DOM = new JSDOM('<!doctype html><html><body></body></html>');

/**
 * Sets up the global vars required for creating DOM elements
 */
function setDomVars() {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  global.window = DOM.window as unknown as typeof global.window;
  global.document = DOM.window.document;
  global.navigator = DOM.window.navigator;
}

setDomVars();

/**
 * @param element A DOM element to snapshot
 * @returns The HTML string representation of the element
 */
export function serializeElement(element: JSX.Element | (() => JSX.Element)): string {
  setDomVars();

  return typeof element === 'function' ? element().outerHTML : element.outerHTML;
}

/**
 * @param element A DOM element to snapshot
 * @returns The HTML string representation of the element
 */
export function serializeElementSandbox(element: () => JSX.Element): string {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  global.window = DOM.window as unknown as typeof global.window;
  global.document = DOM.window.document;
  global.navigator = DOM.window.navigator;

  const snapshot: string = element().outerHTML;

  global.window = WINDOW;
  global.document = DOCUMENT;
  global.navigator = NAVIGATOR;

  return snapshot;
}
