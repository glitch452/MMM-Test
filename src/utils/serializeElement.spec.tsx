import { expect } from 'chai';
import React from 'jsx-dom-cjs';
import { serializeElement, serializeElementSandbox } from './serializeElement';
import 'chai-snapshot-matcher';

describe('serializeElement', () => {
  describe('serializeElement', () => {
    it('should serialize a basic element properly', function () {
      const actual = serializeElement(<div />);
      expect(actual).to.matchSnapshot(this);
    });

    it('should serialize an element with properties properly', function () {
      const actual = serializeElement(<div className="css-class" />);
      expect(actual).to.matchSnapshot(this);
    });

    it('should serialize a basic element properly when passed in as a function', function () {
      const actual = serializeElement(() => <div />);
      expect(actual).to.matchSnapshot(this);
    });
  });

  describe('serializeElementSandbox', () => {
    it('should serialize a basic element properly', function () {
      const actual = serializeElementSandbox(() => <div />);
      expect(actual).to.matchSnapshot(this);
    });

    it('should serialize an element with properties properly', function () {
      const actual = serializeElementSandbox(() => <div className="css-class" />);
      expect(actual).to.matchSnapshot(this);
    });
  });
});
