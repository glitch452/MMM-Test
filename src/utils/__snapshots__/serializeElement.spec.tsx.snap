exports["serializeElement serializeElement should serialize a basic element properly"] = "<div></div>";

exports["serializeElement serializeElement should serialize an element with properties properly"] = "<div class=\"css-class\"></div>";

exports["serializeElement serializeElement should serialize a basic element properly when passed in as a function"] = "<div></div>";

exports["serializeElement serializeElementSandbox should serialize a basic element properly"] = "<div></div>";

exports["serializeElement serializeElementSandbox should serialize an element with properties properly"] = "<div class=\"css-class\"></div>";

